src/types/path.rs

```diff
@@ -0,0 +1,66 @@
+use crate::{FromPyObject, IntoPy, PyAny, PyObject, PyResult, Python, ToPyObject};
+use std::borrow::Cow;
+#[cfg(not(windows))]
+use std::ffi::OsStr;
+use std::ffi::OsString;
+use std::path::{Path, PathBuf};
+
+impl ToPyObject for Path {
+    fn to_object(&self, py: Python) -> PyObject {
+        self.as_os_str().to_object(py)
+    }
+}
+
+// See osstr.rs for why this is Unix-only
+// HELP: should this implementation be here at all? It may be confusing for PyO3 users when a trait
+// is implemented only on some platforms
+#[cfg(not(windows))]
+impl<'a> FromPyObject<'a> for &'a Path {
+    fn extract(ob: &'a PyAny) -> PyResult<Self> {
+        Ok(Path::new(<&OsStr>::extract(ob)?))
+    }
+}
+
+impl FromPyObject<'_> for PathBuf {
+    fn extract(ob: &PyAny) -> PyResult<Self> {
+        Ok(PathBuf::from(OsString::extract(ob)?))
+    }
+}
+
+impl<'a> IntoPy<PyObject> for &'a Path {
+    #[inline]
+    fn into_py(self, py: Python) -> PyObject {
+        self.as_os_str().to_object(py)
+    }
+}
+
+impl<'a> ToPyObject for Cow<'a, Path> {
+    #[inline]
+    fn to_object(&self, py: Python) -> PyObject {
+        self.as_os_str().to_object(py)
+    }
+}
+
+impl ToPyObject for PathBuf {
+    #[inline]
+    fn to_object(&self, py: Python) -> PyObject {
+        self.as_os_str().to_object(py)
+    }
+}
+
+impl IntoPy<PyObject> for PathBuf {
+    fn into_py(self, py: Python) -> PyObject {
+        self.into_os_string().to_object(py)
+    }
+}
+
+impl<'a> IntoPy<PyObject> for &'a PathBuf {
+    #[inline]
+    fn into_py(self, py: Python) -> PyObject {
+        self.as_os_str().to_object(py)
+    }
+}
+
+// These trait implementations simply forward to the OsStr/OsString trait implementations, so I
+// haven't put the effort to test all of this
+// HELP: should I have?
```

Comment:
I appreciate it's a little tedious but would prefer to have some tests here too! You never know when I might come along and accidentally break things 🙃

Reply:
Yes, perhaps an integration test with an accompanied folder of weird file names (think emojis, long filenames, maybe invalid filenames etc.), and the integration test will have both Python and Rust read those filenames, transmit the filenames to the other language respectively, and assert_eq.